### DISCUSSION TITLE
How to draw a dotted line?

### DISCUSSION THREAD

**Discussion by 0xced** [SixLabors/ImageSharp.Drawing#155] [Q&A] [answered]
I'm trying to port some code currently using `System.Drawing.Common` to ImageSharp.Drawing since it was announced that [`System.Drawing.Common` will become Windows-only in .NET 7][1].

I want to draw a dotted (not dashed) line. The code based on `System.Drawing.Common` [sets round caps][2] on the pen but I could not find an equivalent property on ImageSharp's `Pen`.

My current attempt (without round caps, using a pattern of `new[] { 0.1f, 2f }`):
<img width="75" alt="image" src="https://user-images.githubusercontent.com/51363/127771694-7d47de3f-4aa1-43fd-b39f-59457ccb4da1.png">

What it should eventually look like:
<img width="75" alt="image" src="https://user-images.githubusercontent.com/51363/127771702-a3327cb0-0006-44f9-8afb-cae3f25e01d2.png">

Is there an easy way to draw a dotted line with ImageSharp.Drawing?

[1]: https://github.com/dotnet/designs/blob/main/accepted/2021/system-drawing-win-only/system-drawing-win-only.md
[2]: https://github.com/manuelbl/SwissQRBill.NET/blob/0ed8e001ac26712490adba79e2c68937e91f2763/Generator/Canvas/PNGCanvas.cs#L266-L269

**Comment by antonfirsov** [comment, marked as answer]
Personally, I'm not super familiar with the API, but if I'm getting it right, our current [Pen](https://github.com/SixLabors/ImageSharp.Drawing/blob/c29e5de8bb1ed0afd11d71cb75ede5fd4bee0644/src/ImageSharp.Drawing/Processing/Pen.cs) structure doesn't support custom caps, however [this method](https://github.com/SixLabors/ImageSharp.Drawing/blob/0849d85a84584d717924cb7aae3adda8b74743c8/src/ImageSharp.Drawing/Shapes/OutlinePathExtensions.cs#L75) may actually support it. It generates a shape from an line path, so you can draw it with one of the `.Fill()` overloads. [This example](https://github.com/SixLabors/ImageSharp.Drawing/blob/c29e5de8bb1ed0afd11d71cb75ede5fd4bee0644/samples/DrawShapesWithImageSharp/Program.cs) might be also interesting. Can you give it a try?

If it works out, I think it would make sense to expose `JointStyle` and `EndCapStype` on the `IPen` API.

/cc @tocsoft

**Reply by JimBobSquarePants** [reply]
That samples project needs a massive rewrite to bring it up to the current API layout. We should also avoid adding extensions to unrelated types. e.g. `SaveImage(this IPathCollection collection ....` as they do not demonstrate well the natural use of the API and give us a false impression of usability.

**Reply by 0xced** [reply]
The `GenerateOutline` method was exactly what I was looking for. After tweaking the width and pattern parameters I was able to draw a nice dotted line, thanks a lot!

**Reply by antonfirsov** [reply]
@0xced if you feel like contributing back: #158 ;)

**Comment by JimBobSquarePants** [comment]
@0xced I've just pushed a PR to the repository to update the sample code. https://github.com/SixLabors/ImageSharp.Drawing/pull/157

The `Path` extensions method `GenerateOutline` gives you an API that allows choosing the `EndCapStyle`

https://github.com/SixLabors/ImageSharp.Drawing/blob/3ce36e0ac80036c9cb60ba6c12f99cce001f71dd/samples/DrawShapesWithImageSharp/Program.cs#L196-L204

Here's example output with `EndCapStyle.Round`.

![StarOutlineDashed_5_Round_Round](https://user-images.githubusercontent.com/385879/127868077-957904f2-aefd-4a1a-8c2b-780df6ab5117.png)
